### PR TITLE
Add Custom Resource support to "kubectl autoscale"

### LINKED ISSUE #72646: Implement "kubectl autoscale" for Custom Resources
**What would you like to be added**:

I want to create HPA instances via `kubectl autoscale` for our custom controllers which implement the scale subresource, so that I can run

```bash
$ kubectl autoscale vmirs vmi-replicaset-cirros --min=2 --max=5
```

**Why is this needed**:

We are implementing the scale subresource for our CRDs and I can create HPA instances for our controller instances via the declarative workflow (`kubectl create`). However, if I try to create it via `kubectl autoscale`, I get:

```bash
$ kubectl autoscale vmirs vmi-replicaset-cirros --min=2 --max=5
error: no kind "VirtualMachineInstanceReplicaSet" is registered for version "kubevirt.io/v1alpha2"
```
It will make HPA usage for custom controllers more accessible for the imperative workflows.

### PR DESCRIPTION
**Pull request by rmohr**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Generalize the autoscsale command to simply let the dynamic client check
if a scale subresource is registered for the supplied type. This allows
using the autoscale command for built in types as well as for custom
resources.

**Which issue(s) this PR fixes**:

Fixes #72646 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Add Custom Resource support to "kubectl autoscale"
```

### CODE REVIEW

**Comment by rmohr** [comment]
/cc @nikhita @sttts

would you be so kind to have a look on this one?

File: pkg/kubectl/polymorphichelpers/canbeautoscaled_test.go [hunk 1]
```diff
@@ -17,51 +17,105 @@ limitations under the License.
...
+	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/cli-runtime/pkg/genericclioptions/resource"
+	scalefake "k8s.io/client-go/scale/fake"
+	testing2 "k8s.io/client-go/testing"
+	"k8s.io/kubernetes/pkg/kubectl"
 )
 
-func TestCanBeAutoscaled(t *testing.T) {
+func TestCoreTypesCanBeAutoscaled(t *testing.T) {
```
Comment (Rajat-0) [inline pkg/kubectl/polymorphichelpers/canbeautoscaled_test.go]: why do we need to change the function name here?

File: pkg/kubectl/polymorphichelpers/canbeautoscaled.go [hunk 2]
```diff
@@ -36,7 +39,11 @@ func canBeAutoscaled(kind schema.GroupKind) error {
 		extensionsv1beta1.SchemeGroupVersion.WithKind("ReplicaSet").GroupKind():
 		// nothing to do here
 	default:
-		return fmt.Errorf("cannot autoscale a %v", kind)
+		gr := mapping.GroupVersionKind.GroupVersion().WithResource(mapping.Resource.Resource).GroupResource()
+		_, err := scaler.Scales(info.Namespace).Get(gr, info.Name)
```
Comment (sttts) [inline pkg/kubectl/polymorphichelpers/canbeautoscaled.go]: Discovery information has this as well. Can't we drop all the special case above and just depend on that?

Comment (sttts) [reply pkg/kubectl/polymorphichelpers/canbeautoscaled.go]: /cc @s-urbaniak @DirectXMan12

Comment (DirectXMan12) [reply pkg/kubectl/polymorphichelpers/canbeautoscaled.go]: yes, we should be able to

File: pkg/kubectl/polymorphichelpers/canbeautoscaled_test.go (hunk 1 above)
Comment (Rajat-0) [reply pkg/kubectl/polymorphichelpers/canbeautoscaled_test.go]: My doubt was that test case name  according to Golang conventions should be TestFunctionName

File: pkg/kubectl/polymorphichelpers/canbeautoscaled.go (hunk 2 above)
Comment (rmohr) [reply pkg/kubectl/polymorphichelpers/canbeautoscaled.go]: Switched to the dynamic client. Does it look like what you had in mind?

**Comment by rmohr** [comment]
@sttts @DirectXMan12  switched to the dynamic client. Let me know if you had something else in mind.

File: pkg/kubectl/cmd/autoscale/autoscale.go [hunk 3]
```diff
@@ -133,7 +133,11 @@ func (o *AutoscaleOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args
 	o.dryRun = cmdutil.GetFlagBool(cmd, "dry-run")
 	o.createAnnotation = cmdutil.GetFlagBool(cmd, cmdutil.ApplyAnnotationsFlag)
 	o.builder = f.NewBuilder()
-	o.canBeAutoscaled = polymorphichelpers.CanBeAutoscaledFn
+	d, err := f.ToDiscoveryClient()
```
Comment (DirectXMan12) [inline pkg/kubectl/cmd/autoscale/autoscale.go]: please try to avoid one-letter variables.  They're unclear, hard to refactor, and easy to lose track of.

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
minor nit inline, otherwise looks good

File: pkg/kubectl/cmd/autoscale/autoscale.go (hunk 3 above)
Comment (rmohr) [reply pkg/kubectl/cmd/autoscale/autoscale.go]: Done. Thanks.

**Comment by rmohr** [comment]
Let me know if anything else is missing to whitelist it for testing.

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
Thanks for improving this code :+1: and sorry for delay reviewing :blush:

**Pull request merged**